File: csharp/Platform.Data.Doublets.Tests/Uint64LinksExtensionsTests.cs
using Platform.Data.Doublets.Memory;
using Platform.Data.Doublets.Memory.United.Generic;
using Platform.Data.Numbers.Raw;
using Platform.Memory;
using Platform.Numbers;
using Xunit;
using Xunit.Abstractions;
using TLinkAddress = System.UInt64;

namespace Platform.Data.Doublets.Tests
{
    public class Uint64LinksExtensionsTests
    {
        public static ILinks<TLinkAddress> CreateLinks() => CreateLinks<TLinkAddress>(new Platform.IO.TemporaryFile());

        public static ILinks<TLinkAddress> CreateLinks<TLinkAddress>(string dataDBFilename) where TLinkAddress : struct
        {
            var linksConstants = new LinksConstants<TLinkAddress>(enableExternalReferencesSupport: true);
            return new UnitedMemoryLinks<TLinkAddress>(new FileMappedResizableDirectMemory(dataDBFilename), UnitedMemoryLinks<TLinkAddress>.DefaultLinksSizeStep, linksConstants, IndexTreeType.Default);
        }
        [Fact]
        public void FormatStructureWithExternalReferenceTest()
        {
            ILinks<TLinkAddress> links = CreateLinks();
            TLinkAddress zero = default;
            var one = Arithmetic.Increment(zero);
            var markerIndex = one;
            var meaningRoot = links.GetOrCreate(markerIndex, markerIndex);
            var numberMarker = links.GetOrCreate(meaningRoot, Arithmetic.Increment(ref markerIndex));
            AddressToRawNumberConverter<TLinkAddress> addressToNumberConverter = new();
            var numberAddress = addressToNumberConverter.Convert(1);
            var numberLink = links.GetOrCreate(numberMarker, numberAddress);
            var linkNotation = links.FormatStructure(numberLink, link => link.IsFullPoint(), true);
            Assert.Equal("(3:(2:1 2) 18446744073709551615)", linkNotation);
        }
    }
}

File: cpp/Platform.Data.Doublets.Tests/Uint64LinksExtensionsTests.cpp


using TLinkAddress = std::uint64_t;

namespace Platform::Data::Doublets::Tests
{
    TEST_CLASS(Uint64LinksExtensionsTests)
    {
        public: static ILinks<TLinkAddress> CreateLinks() { return CreateLinks<TLinkAddress>(Platform.IO.TemporaryFile(); });

        public: static ILinks<TLinkAddress> CreateLinks<TLinkAddress>(std::string dataDBFilename)
        {
            auto linksConstants = LinksConstants<TLinkAddress>(enableExternalReferencesSupport: true);
            return UnitedMemoryLinks<TLinkAddress>(FileMappedResizableDirectMemory(dataDBFilename), UnitedMemoryLinks<TLinkAddress>.DefaultLinksSizeStep, linksConstants, IndexTreeType.Default);
        }
        public: TEST_METHOD(FormatStructureWithExternalReferenceTest)
        {
            ILinks<TLinkAddress> storage = CreateLinks();
            TLinkAddress zero = 0;
            auto one = zero + 1;
            auto markerIndex = one;
            auto meaningRoot = storage.GetOrCreate(markerIndex, markerIndex);
            auto numberMarker = storage.GetOrCreate(meaningRoot, markerIndex + 1);
            AddressToRawNumberConverter<TLinkAddress> addressToNumberConverter = new();
            auto numberAddress = addressToNumberConverter.Convert(1);
            auto numberLink = storage.GetOrCreate(numberMarker, numberAddress);
            auto linkNotation = storage.FormatStructure(numberLink, link => link.IsFullPoint(), true);
            Assert::AreEqual("(3:(2:1 2) 18446744073709551615)", linkNotation);
        }
    };
}
